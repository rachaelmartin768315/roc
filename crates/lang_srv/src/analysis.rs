use std::{
    collections::{HashMap, HashSet},
    path::{Path, PathBuf},
    sync::Arc,
};

use bumpalo::Bump;
use log::debug;
use roc_can::{abilities::AbilitiesStore, expr::Declarations};
use roc_collections::{MutMap, MutSet};
use roc_load::{CheckedModule, LoadedModule};
use roc_module::symbol::{Interns, ModuleId, Symbol};
use roc_packaging::cache::{self, RocCacheDir};
use roc_region::all::{LineInfo, Region};
use roc_reporting::report::RocDocAllocator;
use roc_solve_problem::TypeError;
use roc_types::{
    subs::{Subs, Variable},
    types::Alias,
};

use tower_lsp::lsp_types::{Diagnostic, SemanticTokenType, Url};

mod analysed_doc;
mod completion;
mod parse_ast;
mod semantic_tokens;
mod tokens;
mod utils;

use crate::convert::diag::{IntoLspDiagnostic, ProblemFmt};

pub(crate) use self::analysed_doc::{AnalyzedDocument, DocInfo};
use self::{analysed_doc::ModuleIdToUrl, tokens::Token};

pub const HIGHLIGHT_TOKENS_LEGEND: &[SemanticTokenType] = Token::LEGEND;

#[derive(Debug, Clone)]
pub(super) struct AnalyzedModule {
    exposed_imports: Vec<(Symbol, Variable)>,
    imports: HashMap<ModuleId, Arc<Vec<(Symbol, Variable)>>>,
    aliases: MutMap<Symbol, (bool, Alias)>,
    module_id: ModuleId,
    interns: Interns,
    subs: Subs,
    abilities: AbilitiesStore,
    declarations: Declarations,
    // We need this because ModuleIds are not stable between compilations, so a ModuleId visible to
    // one module may not be true global to the language server.
    module_id_to_url: ModuleIdToUrl,
}
#[derive(Debug, Clone)]
pub struct AnalysisResult {
    module: Option<AnalyzedModule>,
    diagnostics: Vec<Diagnostic>,
}

pub(crate) fn global_analysis(doc_info: DocInfo) -> Vec<AnalyzedDocument> {
    let fi = doc_info.url.to_file_path().unwrap();
    let src_dir = find_src_dir(&fi).to_path_buf();

    let arena = Bump::new();
    let loaded = roc_load::load_and_typecheck_str(
        &arena,
        fi,
        &doc_info.source,
        src_dir,
        roc_target::TargetInfo::default_x86_64(),
        roc_load::FunctionKind::LambdaSet,
        roc_reporting::report::RenderTarget::Generic,
        RocCacheDir::Persistent(cache::roc_cache_dir().as_path()),
        roc_reporting::report::DEFAULT_PALETTE,
    );

    let module = match loaded {
        Ok(module) => module,
        Err(problem) => {
            let all_problems = problem
                .into_lsp_diagnostic(&())
                .into_iter()
                .collect::<Vec<_>>();

            let analyzed_document = AnalyzedDocument {
                doc_info,
                analysis_result: AnalysisResult {
                    module: None,
                    diagnostics: all_problems,
                },
            };

            return vec![analyzed_document];
        }
    };

    let mut documents = vec![];

    let LoadedModule {
        interns,
        mut can_problems,
        mut type_problems,
        mut declarations_by_id,
        sources,
        mut typechecked,
        solved,
        abilities_store,
        docs_by_module,
        imported_modules,
        mut exposed_imports,
        mut imports,
        exposed_types_storage,
        mut exposes,
        ..
    } = module;

    let mut root_module = Some(RootModule {
        subs: solved.into_inner(),
        abilities_store,
    });
    let exposed_imports: HashMap<_, _> = exposed_imports
        .into_iter()
        .map(|(id, symbols)| {
            (
                id,
                symbols
                    .into_iter()
                    .filter_map(|(symbol, _)| {
                        exposes.get(&id)?.iter().find(|(symb, _)| symb == &symbol)
                    })
                    .cloned()
                    .collect::<Vec<_>>(),
            )
        })
        .collect();
    let exposed: HashMap<_, _> = exposes
        .into_iter()
        .map(|(id, symbols)| (id, Arc::new(symbols)))
        .collect();
    let mut builder = AnalyzedDocumentBuilder {
        interns: &interns,
        module_id_to_url: module_id_to_url_from_sources(&sources),
        can_problems: &mut can_problems,
        type_problems: &mut type_problems,
        declarations_by_id: &mut declarations_by_id,
        typechecked: &mut typechecked,
        root_module: &mut root_module,
        exposed_imports,
        imports: &mut imports,
        exposed,
    };
    debug!("docs: {:#?}", docs_by_module);

    for (module_id, (path, source)) in sources {
        let doc = builder.build_document(path, source, module_id, doc_info.version);
        debug!("================");
        debug!("module:{:?}", module_id);

        // let exposed_imp = exposed_imports.get(&module_id)?;
        if let Some(modu) = &doc.analysis_result.module {
            let aliases = &modu.aliases;
            let imports = &modu.imports;
            // debug!("interns modules:{:#?}", module.interns.module_ids);

            // debug!("exposed_imports:{:#?}", exposed_imp);
            // debug!("exposed:{:#?}", );
            debug!("imports:{:#?}", imports);

            // debug!("docs:{:#?}", docs.1.entries);
            debug!("alais:{:#?}", aliases);
            // debug!("decls:{:#?}", module.declarations)
        }
        documents.push(doc);
    }

    documents
}

fn find_src_dir(path: &Path) -> &Path {
    path.parent().unwrap_or(path)
}

fn _find_parent_git_repo(path: &Path) -> Option<&Path> {
    let mut path = path;
    loop {
        if path.join(".git").exists() {
            return Some(path);
        }

        path = path.parent()?;
    }
}

fn module_id_to_url_from_sources(sources: &MutMap<ModuleId, (PathBuf, Box<str>)>) -> ModuleIdToUrl {
    sources
        .iter()
        .map(|(module_id, (path, _))| {
            let url = path_to_url(path);
            (*module_id, url)
        })
        .collect()
}

fn path_to_url(path: &Path) -> Url {
    if path.is_relative() {
        // Make it <tmpdir>/path
        let tmpdir = std::env::temp_dir();
        Url::from_file_path(tmpdir.join(path)).unwrap()
    } else {
        Url::from_file_path(path).unwrap()
    }
}

struct RootModule {
    subs: Subs,
    abilities_store: AbilitiesStore,
}

struct AnalyzedDocumentBuilder<'a> {
    interns: &'a Interns,
    module_id_to_url: ModuleIdToUrl,
    can_problems: &'a mut MutMap<ModuleId, Vec<roc_problem::can::Problem>>,
    type_problems: &'a mut MutMap<ModuleId, Vec<TypeError>>,
    declarations_by_id: &'a mut MutMap<ModuleId, Declarations>,
    typechecked: &'a mut MutMap<ModuleId, CheckedModule>,
    root_module: &'a mut Option<RootModule>,
    imports: &'a mut MutMap<ModuleId, MutSet<ModuleId>>,
    exposed_imports: HashMap<ModuleId, Vec<(Symbol, Variable)>>,
    exposed: HashMap<ModuleId, Arc<Vec<(Symbol, Variable)>>>,
}

impl<'a> AnalyzedDocumentBuilder<'a> {
    fn build_document(
        &mut self,
        path: PathBuf,
        source: Box<str>,
        module_id: ModuleId,
        version: i32,
    ) -> AnalyzedDocument {
        let subs;
        let abilities;
        let declarations;
        let aliases;
        let imports = self
            .imports
            .remove(&module_id)
            .unwrap_or_default()
            .into_iter()
            .map(|id| {
                (
                    id,
                    self.exposed.get(&id).unwrap_or(&Arc::new(vec![])).clone(),
                )
            })
            .collect::<HashMap<_, _>>();
        let exposed_imports = self.exposed_imports.remove(&module_id).unwrap_or_default();

        if let Some(m) = self.typechecked.remove(&module_id) {
            subs = m.solved_subs.into_inner();
            abilities = m.abilities_store;
            declarations = m.decls;
            aliases = m.aliases;
        } else {
            let rm = self.root_module.take().unwrap();
            subs = rm.subs;
            abilities = rm.abilities_store;
            declarations = self.declarations_by_id.remove(&module_id).unwrap();
            aliases = HashMap::default();
        }

        let analyzed_module = AnalyzedModule {
            exposed_imports,
            imports,
            aliases,
            subs,
            abilities,
            declarations,
            module_id,
            interns: self.interns.clone(),
            module_id_to_url: self.module_id_to_url.clone(),
        };

        let line_info = LineInfo::new(&source);
        let diagnostics = self.build_diagnostics(&path, &source, &line_info, module_id);

        AnalyzedDocument {
            doc_info: DocInfo {
                url: path_to_url(&path),
                line_info,
                source: source.into(),
                version,
            },
            analysis_result: AnalysisResult {
                module: Some(analyzed_module),
                diagnostics,
            },
        }
    }

    fn build_diagnostics(
        &mut self,
        source_path: &Path,
        source: &str,
        line_info: &LineInfo,
        module_id: ModuleId,
    ) -> Vec<Diagnostic> {
        let lines: Vec<_> = source.lines().collect();

        let alloc = RocDocAllocator::new(&lines, module_id, self.interns);

        let mut all_problems = Vec::new();
        let fmt = ProblemFmt {
            alloc: &alloc,
            line_info,
            path: source_path,
        };

        let can_problems = self.can_problems.remove(&module_id).unwrap_or_default();

        let type_problems = self.type_problems.remove(&module_id).unwrap_or_default();

        for can_problem in can_problems {
            if let Some(diag) = can_problem.into_lsp_diagnostic(&fmt) {
                all_problems.push(diag);
            }
        }

        for type_problem in type_problems {
            if let Some(diag) = type_problem.into_lsp_diagnostic(&fmt) {
                all_problems.push(diag);
            }
        }

        all_problems
    }
}
