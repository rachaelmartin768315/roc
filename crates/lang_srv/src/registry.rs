use std::{
    cell::OnceCell,
    collections::HashMap,
    future::Future,
    io::{stderr, Write},
    ops::Deref,
    rc::Rc,
    sync::Arc,
};

use tokio::{
    io::AsyncRead,
    sync::{Mutex, MutexGuard, RwLock, RwLockReadGuard, RwLockWriteGuard},
    task::JoinHandle,
};
use tower_lsp::lsp_types::{
    notification::Notification, CompletionResponse, Diagnostic, GotoDefinitionResponse, Hover,
    Position, SemanticTokensResult, TextEdit, Url,
};

use crate::analysis::{global_anal, AnalysisResult, AnalyzedDocument, DocInfo};

pub(crate) enum DocumentChange {
    Modified(Url, String, u32),
    Closed(Url),
}
#[derive(Debug)]
pub(crate) struct LatestDocument {
    info: DocInfo,
    //We can hold this mutex locked during updating while the latest and doc_info are out of sync
    //the lock should be aquired and immediately freed by and task looking to get a copy of info
    //At the top level we will need to store our lock
    latest: tokio::sync::watch::Receiver<Option<Arc<AnalyzedDocument>>>,
    latest_sender: tokio::sync::watch::Sender<Option<Arc<AnalyzedDocument>>>,
}
impl LatestDocument {
    pub(crate) async fn get_latest(&self) -> Arc<AnalyzedDocument> {
        let mut my_reciever = self.latest.clone();

        let a = my_reciever.wait_for(|x| x.is_some()).await.unwrap();
        match a.as_ref() {
            Some(latest) => latest.clone(),
            None => todo!(),
        }
    }
    pub(crate) fn set_latest(&self, latest: Arc<AnalyzedDocument>) {
        self.latest_sender.send(Some(latest)).unwrap()
    }
    pub(crate) fn waiting_for_doc(&self) -> bool {
        self.latest.borrow().is_none()
    }
    pub(crate) fn new(doc_info: DocInfo) -> LatestDocument {
        let chan = tokio::sync::watch::channel(None);
        LatestDocument {
            info: doc_info,
            latest_sender: chan.0,
            latest: chan.1,
        }
    }
    pub(crate) fn new_initialised(doc: Arc<AnalyzedDocument>) -> LatestDocument {
        let info = doc.doc_info.clone();
        let chan = tokio::sync::watch::channel(Some(doc));
        LatestDocument {
            info,
            latest_sender: chan.0,
            latest: chan.1,
        }
    }
}

#[derive(Debug)]
pub(crate) struct DocumentPair {
    latest_document: LatestDocument,
    last_good_document: Arc<AnalyzedDocument>,
}

// pub(crate) fn new(latest: AnalyzedDocument, last_good: AnalyzedDocument) -> DocumentPair {
//     DocumentPair {
//         //TODO not sure if i should actually be cloning here?
//         latest_document: (latest.doc_info, Arc::new(RwLock::new(last_good.clone()))),
//         last_good_document: last_good,
//     }
// }
// pub(crate) fn new_latest_type_checked(latest_doc: AnalyzedDocument) -> DocumentPair {
//     DocumentPair {
//         //TODO not sure if i should actually be cloning here?
//         latest_document: (
//             latest_doc.doc_info.clone(),
//             Arc::new(RwLock::new(latest_doc.clone())),
//         ),
//         last_good_document: latest_doc,
//     }
// }

#[derive(Debug, Default)]
pub(crate) struct Registry {
    documents: Mutex<HashMap<Url, DocumentPair>>,
}

impl Registry {
    fn update_document<'a>(
        documents: &mut MutexGuard<'a, HashMap<Url, DocumentPair>>,
        document: AnalyzedDocument,
    ) {
        let url = document.url().clone();
        let document = Arc::new(document);
        // writeln!(
        //     std::io::stderr(),
        //     "updating doc{:?}. version:{:?}",
        //     &url,
        //     &document.doc_info.version
        // );

        let latest_doc = LatestDocument::new_initialised(document.clone());
        match documents.get_mut(&url) {
            Some(old_doc) => {
                //This is a special case where we know we should
                if old_doc.latest_document.waiting_for_doc() {
                    old_doc.latest_document.set_latest(document.clone());
                }
                if document.type_checked() {
                    *old_doc = DocumentPair {
                        latest_document: latest_doc,
                        last_good_document: document,
                    };
                } else {
                    //TODO this seems ugly but for now i'll let it slide. shoudl be immutable
                    *old_doc = DocumentPair {
                        latest_document: latest_doc,
                        last_good_document: old_doc.last_good_document.clone(),
                    };
                }
            }
            None => {
                documents.insert(
                    url.clone(),
                    DocumentPair {
                        latest_document: latest_doc,
                        last_good_document: document,
                    },
                );
            }
        }
    }

    pub async fn apply_change(&self, analysed_docs: Vec<AnalyzedDocument>) -> () {
        writeln!(
            std::io::stderr(),
            "updated the latest document with docinfo"
        );

        let mut documents = self.documents.lock().await;
        writeln!(
            std::io::stderr(),
            "finised doc analasys updating docs {:?}",
            analysed_docs
                .iter()
                .map(|a| a.doc_info.url.to_string())
                .collect::<Vec<_>>()
        );

        for document in analysed_docs {
            Registry::update_document(&mut documents, document);
        }
    }

    pub async fn apply_doc_info_changes(&self, url: Url, partial: DocInfo) {
        let mut lock = self.documents.lock().await;
        let doc = lock.get_mut(&url);
        match doc {
            Some(a) => {
                writeln!(
                    std::io::stderr(),
                    "set the docInfo to version:{:?}",
                    partial.version
                );

                a.latest_document = LatestDocument::new(partial);
            }

            None => (),
        }
    }

    fn document_info_by_url(&self, url: &Url) -> Option<DocInfo> {
        self.documents
            .blocking_lock()
            .get(url)
            .map(|a| a.latest_document.info.clone())
    }
    async fn latest_document_by_url(&self, url: &Url) -> Option<Arc<AnalyzedDocument>> {
        match self.documents.lock().await.get(url) {
            Some(a) => Some(a.latest_document.get_latest().await),
            None => None,
        }
    }

    pub async fn diagnostics(&self, url: &Url) -> Vec<Diagnostic> {
        let Some( document) = self.latest_document_by_url(url).await else {
            return vec![];
        };
        document.diagnostics()
    }

    pub async fn hover(&self, url: &Url, position: Position) -> Option<Hover> {
        self.latest_document_by_url(url).await?.hover(position)
    }

    pub async fn goto_definition(
        &self,
        url: &Url,
        position: Position,
    ) -> Option<GotoDefinitionResponse> {
        let document = self.latest_document_by_url(url).await?;
        let symbol = document.symbol_at(position)?;
        let def_document_url = document.module_url(symbol.module_id())?;
        let def_document = self.latest_document_by_url(&def_document_url).await?;
        def_document.definition(symbol)
    }

    pub fn formatting(&self, url: &Url) -> Option<Vec<TextEdit>> {
        let document = self.document_info_by_url(url)?;
        document.format()
    }

    pub fn semantic_tokens(&self, url: &Url) -> Option<SemanticTokensResult> {
        let document = self.document_info_by_url(url)?;
        document.semantic_tokens()
    }
    pub async fn completion_items(
        &self,
        url: &Url,
        position: Position,
    ) -> Option<CompletionResponse> {
        let lock = self.documents.lock().await;
        let pair = lock.get(url)?;
        writeln!(stderr(), "got document");
        let latest_doc_info = &pair.latest_document.info;
        writeln!(stderr(), "latest version:{:?} ", latest_doc_info.version);

        let symbol_prefix = latest_doc_info.get_prefix_at_position(position);

        //this strategy probably won't work for record fields
        let completions =
            pair.last_good_document
                .completion_items(position, &latest_doc_info, symbol_prefix)?;

        Some(CompletionResponse::Array(completions))
    }
}
