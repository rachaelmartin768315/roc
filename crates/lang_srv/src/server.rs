use analysis::HIGHLIGHT_TOKENS_LEGEND;
use log::{debug, trace};
use registry::Registry;
use std::future::Future;

use std::time::Duration;
use tower_lsp::jsonrpc::Result;
use tower_lsp::lsp_types::*;
use tower_lsp::{Client, LanguageServer, LspService, Server};

use crate::analysis::global_analysis;

mod analysis;
mod convert;
mod registry;

#[derive(Debug)]
struct RocLs {
    pub inner: Inner,
    client: Client,
}

///This exists so we can test most of RocLs without anything LSP related
#[derive(Debug)]
struct Inner {
    registry: Registry,
}

impl std::panic::RefUnwindSafe for RocLs {}

impl RocLs {
    pub fn new(client: Client) -> Self {
        Self {
            inner: Inner::new(),
            client,
        }
    }
    ///Wait for all the semaphores associated with an in-progress document_info update to be released
    pub fn capabilities() -> ServerCapabilities {
        let text_document_sync = TextDocumentSyncCapability::Options(
            // TODO: later on make this incremental
            TextDocumentSyncOptions {
                open_close: Some(true),
                change: Some(TextDocumentSyncKind::FULL),
                ..TextDocumentSyncOptions::default()
            },
        );
        let hover_provider = HoverProviderCapability::Simple(true);
        let definition_provider = DefinitionOptions {
            work_done_progress_options: WorkDoneProgressOptions {
                work_done_progress: None,
            },
        };
        let document_formatting_provider = DocumentFormattingOptions {
            work_done_progress_options: WorkDoneProgressOptions {
                work_done_progress: None,
            },
        };
        let semantic_tokens_provider =
            SemanticTokensServerCapabilities::SemanticTokensOptions(SemanticTokensOptions {
                work_done_progress_options: WorkDoneProgressOptions {
                    work_done_progress: None,
                },
                legend: SemanticTokensLegend {
                    token_types: HIGHLIGHT_TOKENS_LEGEND.into(),
                    token_modifiers: vec![],
                },
                range: None,
                full: Some(SemanticTokensFullOptions::Bool(true)),
            });
        let completion_provider = CompletionOptions {
            resolve_provider: Some(false),
            trigger_characters: Some(vec![".".to_string()]),
            //TODO: what is this?
            all_commit_characters: None,
            work_done_progress_options: WorkDoneProgressOptions {
                work_done_progress: None,
            },
        };
        ServerCapabilities {
            text_document_sync: Some(text_document_sync),
            hover_provider: Some(hover_provider),
            definition_provider: Some(OneOf::Right(definition_provider)),
            document_formatting_provider: Some(OneOf::Right(document_formatting_provider)),
            semantic_tokens_provider: Some(semantic_tokens_provider),
            completion_provider: Some(completion_provider),
            ..ServerCapabilities::default()
        }
    }

    /// Records a document content change.
    async fn change(&self, fi: Url, text: String, version: i32) {
        let updating_result = self.inner.change(&fi, text, version).await;

        //The analysis task can be cancelled by another change coming in which will update the watched variable
        if let Err(e) = updating_result {
            debug!("cancelled change. Reason:{:?}", e);
            return;
        }
        debug!("applied_change getting and returning diagnostics");

        let diagnostics = self.inner.registry.diagnostics(&fi).await;

        self.client
            .publish_diagnostics(fi, diagnostics, Some(version))
            .await;
    }
}

impl Inner {
    pub fn new() -> Inner {
        Self {
            registry: Registry::default(),
        }
    }

    async fn registry(&self) -> &Registry {
        &self.registry
    }

    async fn close(&self, _fi: Url) {
        ()
    }

    pub async fn change(
        &self,
        fi: &Url,
        text: String,
        version: i32,
    ) -> std::result::Result<(), String> {
        debug!("V{:?}:starting change", version);
        //was write lock

        debug!("V{:?}:change acquired registry lock", version);
        let (results, partial) = global_analysis(fi.clone(), text, version);

        self.registry
            .apply_doc_info_changes(fi.clone(), partial.clone())
            .await;
        //Now that we've got our new partial document written and we hold the exclusive write_handle to its analysis we can allow other tasks to access the registry and the doc_info inside this partial document

        debug!(
            "V{:?}:finished updating docinfo, starting analysis ",
            version
        );

        let inner_ref = self;
        let updating_result = async {
            //This reduces wasted computation by waiting to see if a new change comes in, but does delay the final analysis. Ideally this would be replaced with cancelling the analysis when a new one comes in.
            tokio::time::sleep(Duration::from_millis(100)).await;
            let is_latest = inner_ref
                .registry
                .get_latest_version(fi)
                .await
                .map(|latest| latest == version)
                .unwrap_or(true);
            if !is_latest {
                return Err("Not latest version skipping analysis".to_string());
            }

            let results = match tokio::task::spawn_blocking(results).await {
                Err(e) => return Err(format!("Document analysis failed. reason:{:?}", e)),
                Ok(a) => a,
            };
            let latest_version = inner_ref.registry.get_latest_version(fi).await;

            //if this version is not the latest another change must have come in and this analysis is useless
            //if there is no older version we can just proceed with the update
            if let Some(latest_version) = latest_version {
                if latest_version != version {
                    return Err(format!(
                        "version {0} doesn't match latest: {1} discarding analysis  ",
                        version, latest_version
                    ));
                }
            }
            debug!(
                "V{:?}:finished document analysis applying changes ",
                version
            );

            inner_ref.registry.apply_changes(results, fi.clone()).await;
            Ok(())
        }
        .await;
        debug!("V{:?}:finished document change process", version);
        updating_result
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for RocLs {
    async fn initialize(&self, _: InitializeParams) -> Result<InitializeResult> {
        Ok(InitializeResult {
            capabilities: Self::capabilities(),
            ..InitializeResult::default()
        })
    }

    async fn initialized(&self, _: InitializedParams) {
        self.client
            .log_message(MessageType::INFO, "Roc language server initialized.")
            .await;
    }

    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        let TextDocumentItem {
            uri, text, version, ..
        } = params.text_document;
        self.change(uri, text, version).await;
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        let VersionedTextDocumentIdentifier { uri, version, .. } = params.text_document;

        // NOTE: We specify that we expect full-content syncs in the server capabilities,
        // so here we assume the only change passed is a change of the entire document's content.
        let TextDocumentContentChangeEvent { text, .. } =
            params.content_changes.into_iter().next().unwrap();

        trace!("got did_change");
        self.change(uri, text, version).await;
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        let TextDocumentIdentifier { uri } = params.text_document;
        self.inner.close(uri).await;
    }

    async fn shutdown(&self) -> Result<()> {
        Ok(())
    }

    async fn hover(&self, params: HoverParams) -> Result<Option<Hover>> {
        let HoverParams {
            text_document_position_params:
                TextDocumentPositionParams {
                    text_document,
                    position,
                },
            work_done_progress_params: _,
        } = params;

        panic_wrapper_async(|| async {
            self.inner
                .registry
                .hover(&text_document.uri, position)
                .await
        })
        .await
    }

    async fn goto_definition(
        &self,
        params: GotoDefinitionParams,
    ) -> Result<Option<GotoDefinitionResponse>> {
        let GotoDefinitionParams {
            text_document_position_params:
                TextDocumentPositionParams {
                    text_document,
                    position,
                },
            work_done_progress_params: _,
            partial_result_params: _,
        } = params;

        panic_wrapper_async(|| async {
            self.inner
                .registry()
                .await
                .goto_definition(&text_document.uri, position)
                .await
        })
        .await
    }

    async fn formatting(&self, params: DocumentFormattingParams) -> Result<Option<Vec<TextEdit>>> {
        let DocumentFormattingParams {
            text_document,
            options: _,
            work_done_progress_params: _,
        } = params;

        panic_wrapper_async(|| async {
            self.inner
                .registry()
                .await
                .formatting(&text_document.uri)
                .await
        })
        .await
    }

    async fn semantic_tokens_full(
        &self,
        params: SemanticTokensParams,
    ) -> Result<Option<SemanticTokensResult>> {
        let SemanticTokensParams {
            text_document,
            work_done_progress_params: _,
            partial_result_params: _,
        } = params;

        panic_wrapper_async(|| async {
            self.inner
                .registry()
                .await
                .semantic_tokens(&text_document.uri)
                .await
        })
        .await
    }
    async fn completion(&self, params: CompletionParams) -> Result<Option<CompletionResponse>> {
        let doc = params.text_document_position;
        trace!("got completion request");
        let res = panic_wrapper_async(|| async {
            self.inner
                .registry
                .completion_items(&doc.text_document.uri, doc.position)
                .await
        })
        .await;
        res
    }
}

async fn panic_wrapper_async<Fut, T>(
    f: impl FnOnce() -> Fut + std::panic::UnwindSafe,
) -> Result<Option<T>>
where
    Fut: Future<Output = Option<T>>,
{
    match std::panic::catch_unwind(f) {
        Ok(r) => Ok(r.await),
        Err(_) => Err(tower_lsp::jsonrpc::Error::internal_error()),
    }
}

#[tokio::main(flavor = "multi_thread")]
async fn main() {
    env_logger::Builder::from_env("ROCLS_LOG").init();

    let stdin = tokio::io::stdin();
    let stdout = tokio::io::stdout();

    let (service, socket) = LspService::new(RocLs::new);
    Server::new(stdin, stdout, socket).serve(service).await;
}

#[cfg(test)]
mod tests {
    use std::{
        sync::{Once, OnceLock},
        time::Duration,
    };

    use insta::assert_debug_snapshot;
    use tokio::{join, spawn};

    use super::*;
    const DOC_LIT: &str = r#"
app "fizz-buzz"
    packages { pf: "https://github.com/roc-lang/basic-cli/releases/download/0.5.0/Cufzl36_SnJ4QbOoEmiJ5dIpUxBvdB3NEySvuH82Wio.tar.br" }
    imports [pf.Stdout,pf.Task.{ Task, await },]
    provides [main] to pf
"#;
    static INIT: Once = Once::new();
    async fn test_setup(doc: String) -> (Inner, Url) {
        INIT.call_once(|| {
            env_logger::builder()
                .is_test(true)
                .filter_level(log::LevelFilter::Trace)
                .init();
        });
        // static INNER_CELL: OnceLock<Inner> = OnceLock::new();
        // INNER_CELL.set(Inner::new()).unwrap();
        // static URL_CELL: OnceLock<Url> = OnceLock::new();
        // URL_CELL.set().unwrap()).unwrap();

        // let inner = INNER_CELL.get().unwrap();
        let url = Url::parse("file:/test.roc").unwrap();

        let inner = Inner::new();
        //setup the file
        inner.change(&url, doc, 0).await.unwrap();
        (inner, url)
    }

    #[tokio::test]
    async fn test_completion_with_changes() {
        let doc = DOC_LIT.to_string()
            + r#"rec=\a,b->{one:{potato:\d->d,leak:59},two:b}
rectest= 
  value= rec 1 2
  va"#;
        let (inner, url) = test_setup(doc.clone()).await;
        static INNER_CELL: OnceLock<Inner> = OnceLock::new();
        INNER_CELL.set(inner).unwrap();
        static URL_CELL: OnceLock<Url> = OnceLock::new();
        URL_CELL.set(url).unwrap();

        let inner = INNER_CELL.get().unwrap();
        let url = URL_CELL.get().unwrap();
        let position = Position::new(8, 8);
        //setup the file
        inner.change(&url, doc.clone(), 1).await.unwrap();

        //apply a sequence of changes back to back
        let a1 = spawn(inner.change(&url, doc.clone() + "l", 2));
        let a2 = spawn(inner.change(&url, doc.clone() + "lu", 3));
        let a3 = spawn(inner.change(&url, doc.clone() + "lue", 4));
        let a4 = spawn(inner.change(&url, doc.clone() + "lue.", 5));
        //start a completion that would only work if all changes have been applied
        let comp = spawn(async move {
            let url2 = url.clone();
            let reg = inner.registry().await;
            reg.completion_items(&url2, position).await
        });
        // Simulate two changes coming in with a slight delay
        let a = spawn(inner.change(&url, doc.clone() + "lue.o", 6));
        tokio::time::sleep(Duration::from_millis(100)).await;
        let rest = spawn(inner.change(&url, doc.clone() + "lue.on", 7));

        let done = join!(a1, a2, a3, a4, comp, a, rest);

        assert_debug_snapshot!(done)
    }
    #[tokio::test]
    async fn test_completion_as_identifier() {
        let suffix = DOC_LIT.to_string()
            + r#"
main =
  when a is
    inn as outer -> "#;
        let (inner, url) = test_setup(suffix.clone()).await;
        //test compltion for outer
        let position = Position::new(8, 21);

        let change = suffix.clone() + "o";
        inner.change(&url, change, 1).await.unwrap();
        let comp1 = inner
            .registry()
            .await
            .completion_items(&url, position)
            .await;

        let c = suffix.clone() + "i";
        inner.change(&url, c, 2).await.unwrap();
        let comp2 = inner
            .registry()
            .await
            .completion_items(&url, position)
            .await;

        let actual = [comp1, comp2];
        assert_debug_snapshot!(actual)
    }

    #[tokio::test]
    async fn test_completion_as_record() {
        let doc = DOC_LIT.to_string()
            + r#"
main =
  when a is
    {one,two} as outer -> "#;
        let (inner, url) = test_setup(doc.clone()).await;
        //test compltion for outer
        let position = Position::new(8, 27);

        let change = doc.clone() + "o";
        inner.change(&url, change, 1).await.unwrap();
        let comp1 = inner
            .registry()
            .await
            .completion_items(&url, position)
            .await;

        let c = doc.clone() + "t";
        inner.change(&url, c, 2).await.unwrap();
        let comp2 = inner
            .registry()
            .await
            .completion_items(&url, position)
            .await;

        let actual = [comp1, comp2];
        assert_debug_snapshot!(actual);
    }
}
