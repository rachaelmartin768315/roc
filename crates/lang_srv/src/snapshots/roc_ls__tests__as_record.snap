---
source: crates/lang_srv/src/server.rs
expression: actual
---
[
    Some(
        Array(
            [
                CompletionItem {
                    label: "one",
                    kind: Some(
                        Variable,
                    ),
                    detail: Some(
                        "*",
                    ),
                    documentation: None,
                    deprecated: None,
                    preselect: None,
                    sort_text: None,
                    filter_text: None,
                    insert_text: None,
                    insert_text_format: None,
                    insert_text_mode: None,
                    text_edit: None,
                    additional_text_edits: None,
                    command: None,
                    commit_characters: None,
                    data: None,
                    tags: None,
                },
                CompletionItem {
                    label: "two",
                    kind: Some(
                        Variable,
                    ),
                    detail: Some(
                        "*",
                    ),
                    documentation: None,
                    deprecated: None,
                    preselect: None,
                    sort_text: None,
                    filter_text: None,
                    insert_text: None,
                    insert_text_format: None,
                    insert_text_mode: None,
                    text_edit: None,
                    additional_text_edits: None,
                    command: None,
                    commit_characters: None,
                    data: None,
                    tags: None,
                },
                CompletionItem {
                    label: "outer",
                    kind: Some(
                        Variable,
                    ),
                    detail: Some(
                        "*",
                    ),
                    documentation: None,
                    deprecated: None,
                    preselect: None,
                    sort_text: None,
                    filter_text: None,
                    insert_text: None,
                    insert_text_format: None,
                    insert_text_mode: None,
                    text_edit: None,
                    additional_text_edits: None,
                    command: None,
                    commit_characters: None,
                    data: None,
                    tags: None,
                },
            ],
        ),
    ),
    Some(
        Array(
            [
                CompletionItem {
                    label: "one",
                    kind: Some(
                        Variable,
                    ),
                    detail: Some(
                        "*",
                    ),
                    documentation: None,
                    deprecated: None,
                    preselect: None,
                    sort_text: None,
                    filter_text: None,
                    insert_text: None,
                    insert_text_format: None,
                    insert_text_mode: None,
                    text_edit: None,
                    additional_text_edits: None,
                    command: None,
                    commit_characters: None,
                    data: None,
                    tags: None,
                },
                CompletionItem {
                    label: "two",
                    kind: Some(
                        Variable,
                    ),
                    detail: Some(
                        "*",
                    ),
                    documentation: None,
                    deprecated: None,
                    preselect: None,
                    sort_text: None,
                    filter_text: None,
                    insert_text: None,
                    insert_text_format: None,
                    insert_text_mode: None,
                    text_edit: None,
                    additional_text_edits: None,
                    command: None,
                    commit_characters: None,
                    data: None,
                    tags: None,
                },
                CompletionItem {
                    label: "outer",
                    kind: Some(
                        Variable,
                    ),
                    detail: Some(
                        "*",
                    ),
                    documentation: None,
                    deprecated: None,
                    preselect: None,
                    sort_text: None,
                    filter_text: None,
                    insert_text: None,
                    insert_text_format: None,
                    insert_text_mode: None,
                    text_edit: None,
                    additional_text_edits: None,
                    command: None,
                    commit_characters: None,
                    data: None,
                    tags: None,
                },
            ],
        ),
    ),
]
